File: api/predict.js
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";

const model = JSON.parse(readFileSync(new URL("../outputs/models/poor_sleep_preoutcome_model.json", import.meta.url), "utf8"));
const requests = new Map();

function probability(row) {
  return model.estimators.reduce((total, estimator) => {
    const values = estimator.numeric.map((field) => ((Number(row[field.name] ?? field.impute) - field.mean) / field.scale));
    for (const field of estimator.categorical) {
      const value = String(row[field.name] ?? field.impute);
      values.push(...field.categories.map((category) => value === category ? 1 : 0));
    }
    const score = estimator.intercept + values.reduce((sum, value, index) => sum + value * estimator.coefficients[index], 0);
    const calibrated = 1 / (1 + Math.exp(estimator.calibration.a * score + estimator.calibration.b));
    return total + calibrated;
  }, 0) / model.estimators.length;
}

export default function handler(request, response) {
  const started = Date.now();
  response.setHeader("Cache-Control", "no-store");
  response.setHeader("X-Content-Type-Options", "nosniff");
  if (request.method !== "POST") return response.status(405).json({ error: "method_not_allowed" });
  if (process.env.PREDICTION_API_KEY && request.headers["x-api-key"] !== process.env.PREDICTION_API_KEY) return response.status(401).json({ error: "unauthorized" });
  const client = String(request.headers["x-forwarded-for"] ?? "local").split(",")[0];
  const now = Date.now();
  const bucket = (requests.get(client) ?? []).filter((timestamp) => now - timestamp < 60_000);
  if (bucket.length >= 30) return response.status(429).json({ error: "rate_limit", retry_after_seconds: 60 });
  bucket.push(now); requests.set(client, bucket);
  try {
    const supplied = request.body;
    if (!supplied || typeof supplied !== "object" || Array.isArray(supplied)) return response.status(400).json({ error: "object_required" });
    const unknown = Object.keys(supplied).filter((key) => !model.features.includes(key));
    if (unknown.length) return response.status(400).json({ error: "unknown_features", fields: unknown.sort() });
    const row = { ...model.defaults }; const warnings = [];
    for (const [key, raw] of Object.entries(supplied)) {
      if (model.ranges[key]) {
        const value = Number(raw); const bounds = model.ranges[key];
        if (!Number.isFinite(value)) return response.status(422).json({ error: "number_required", field: key });
        if (value < bounds.min || value > bounds.max) return response.status(422).json({ error: "out_of_range", field: key, bounds });
        const baseline = model.drift_baseline.numeric[key];
        if (baseline.sd && Math.abs(value - baseline.mean) / baseline.sd > 3) warnings.push(`Input distribution warning for ${key}`);
        row[key] = value;
      } else {
        const value = String(raw);
        if (!model.categories[key]?.includes(value)) warnings.push(`Unseen category for ${key}`);
        row[key] = value;
      }
    }
    const risk = probability(row); const riskBand = risk >= model.threshold ? "higher" : "lower";
    console.log(JSON.stringify({ level: "info", event: "prediction", model_version: model.version,
      input_hash: createHash("sha256").update(JSON.stringify(supplied)).digest("hex").slice(0, 16), risk_band: riskBand,
      warning_count: warnings.length, duration_ms: Date.now() - started }));
    return response.status(200).json({ probability: risk, classification: Number(risk >= model.threshold), risk_band: riskBand,
      threshold: model.threshold, model_version: model.version, validation_status: "internal synthetic-data validation only", warnings });
  } catch (error) {
    console.error(JSON.stringify({ level: "error", event: "prediction_failed", error: error?.name ?? "Error", duration_ms: Date.now() - started }));
    return response.status(400).json({ error: "invalid_request" });
  }
}

File: api/predict.py
import hashlib
import json
import os
import time
from collections import defaultdict, deque
from http.server import BaseHTTPRequestHandler
from pathlib import Path

import joblib
import pandas as pd

ARTIFACT = Path(__file__).resolve().parents[1] / "outputs/models/poor_sleep_preoutcome_model.joblib"
BUNDLE = joblib.load(ARTIFACT)
REQUESTS = defaultdict(deque)


class handler(BaseHTTPRequestHandler):
    def _send(self, status, payload):
        body = json.dumps(payload).encode()
        self.send_response(status); self.send_header("Content-Type", "application/json")
        self.send_header("Cache-Control", "no-store"); self.send_header("X-Content-Type-Options", "nosniff")
        self.end_headers(); self.wfile.write(body)

    def do_POST(self):
        started = time.time(); request_id = self.headers.get("x-vercel-id", "local")
        configured_key = os.getenv("PREDICTION_API_KEY")
        if configured_key and self.headers.get("x-api-key") != configured_key:
            return self._send(401, {"error": "unauthorized"})
        client = self.headers.get("x-forwarded-for", "local").split(",")[0]
        now = time.time(); bucket = REQUESTS[client]
        while bucket and now - bucket[0] > 60: bucket.popleft()
        if len(bucket) >= 30: return self._send(429, {"error": "rate_limit", "retry_after_seconds": 60})
        bucket.append(now)
        try:
            length = int(self.headers.get("content-length", "0"))
            if length <= 0 or length > 16384: return self._send(413, {"error": "invalid_payload_size"})
            supplied = json.loads(self.rfile.read(length))
            if not isinstance(supplied, dict): return self._send(400, {"error": "object_required"})
            unknown = sorted(set(supplied) - set(BUNDLE["features"]))
            if unknown: return self._send(400, {"error": "unknown_features", "fields": unknown})
            row, warnings = dict(BUNDLE["defaults"]), []
            for key, value in supplied.items():
                if key in BUNDLE["ranges"]:
                    value = float(value); bounds = BUNDLE["ranges"][key]
                    if value < bounds["min"] or value > bounds["max"]:
                        return self._send(422, {"error": "out_of_range", "field": key, "bounds": bounds})
                    baseline = BUNDLE.get("drift_baseline", {}).get("numeric", {}).get(key)
                    if baseline and baseline["sd"] and abs(value - baseline["mean"]) / baseline["sd"] > 3:
                        warnings.append(f"Input distribution warning for {key}")
                elif str(value) not in BUNDLE["categories"].get(key, []):
                    warnings.append(f"Unseen category for {key}")
                row[key] = value
            probability = float(BUNDLE["model"].predict_proba(pd.DataFrame([row]))[0, 1])
            threshold = float(BUNDLE["threshold"])
            payload = {"probability": probability, "classification": int(probability >= threshold),
                       "risk_band": "higher" if probability >= threshold else "lower",
                       "threshold": threshold, "model_version": BUNDLE["version"],
                       "validation_status": "internal synthetic-data validation only", "warnings": warnings}
            audit = {"level": "info", "event": "prediction", "request_id": request_id,
                     "model_version": BUNDLE["version"], "input_hash": hashlib.sha256(json.dumps(supplied, sort_keys=True).encode()).hexdigest()[:16],
                     "risk_band": payload["risk_band"], "warning_count": len(warnings),
                     "duration_ms": round((time.time()-started)*1000)}
            print(json.dumps(audit), flush=True)
            self._send(200, payload)
        except Exception as error:
            print(json.dumps({"level": "error", "event": "prediction_failed", "request_id": request_id,
                              "error": type(error).__name__, "duration_ms": round((time.time()-started)*1000)}), flush=True)
            self._send(400, {"error": "invalid_request"})
